autora/controller/state/history.py

```diff
@@ -0,0 +1,613 @@
+            >>> History()
+            History([])
+
+            ... or with values for any or all of the parameters:
+            >>> from autora.variable import VariableCollection
+            >>> History(metadata=VariableCollection()) # doctest: +ELLIPSIS
+            History([Result(data=VariableCollection(...), kind=ResultKind.METADATA)])
+
+            >>> History(params={"some": "params"})
+            History([Result(data={'some': 'params'}, kind=ResultKind.PARAMS)])
+
+            >>> History(conditions=["a condition"])
+            History([Result(data='a condition', kind=ResultKind.CONDITION)])
+
+            >>> History(observations=["an observation"])
+            History([Result(data='an observation', kind=ResultKind.OBSERVATION)])
+
+            >>> from sklearn.linear_model import LinearRegression
+            >>> History(theories=[LinearRegression()])
+            History([Result(data=LinearRegression(), kind=ResultKind.THEORY)])
+
+            Parameters passed to the constructor are included in the history in the following order:
+            `history`, `metadata`, `params`, `conditions`, `observations`, `theories`
+            >>> History(theories=['t1', 't2'], conditions=['c1', 'c2'],
+            ...     observations=['o1', 'o2'], params={'a': 'param'}, metadata=VariableCollection(),
+            ...     history=[Result("from history", ResultKind.METADATA)]
+            ... )  # doctest: +ELLIPSIS +NORMALIZE_WHITESPACE
+            History([Result(data='from history', kind=ResultKind.METADATA),
+                                    Result(data=VariableCollection(...), kind=ResultKind.METADATA),
+                                    Result(data={'a': 'param'}, kind=ResultKind.PARAMS),
+                                    Result(data='c1', kind=ResultKind.CONDITION),
+                                    Result(data='c2', kind=ResultKind.CONDITION),
+                                    Result(data='o1', kind=ResultKind.OBSERVATION),
+                                    Result(data='o2', kind=ResultKind.OBSERVATION),
+                                    Result(data='t1', kind=ResultKind.THEORY),
+                                    Result(data='t2', kind=ResultKind.THEORY)])
+        """
+        self._history: List
+
+        if history is not None:
+            self._history = list(history)
+        else:
+            self._history = []
+
+        self._history += _init_result_list(
+            metadata=metadata,
+            params=params,
+            conditions=conditions,
+            observations=observations,
+            theories=theories,
+        )
+
+    def update(
+        self,
+        metadata=None,
+        params=None,
+        conditions=None,
+        observations=None,
+        theories=None,
+        history=None,
+    ):
+        """
+        Create a new object with updated values.
+
+        Examples:
+            The initial object is empty:
+            >>> s0 = History()
```

Comment:
This is just minor, but perhaps we should write `h0` instead of `s0` to not confuse it with the Snapshot.

Reply:
Yes, will do that:
- [ ] Update `s` to `h` in history file